core/math/basis.cpp

```diff
@@ -89,13 +89,26 @@ Basis Basis::orthogonalized() const {
 	return c;
 }
 
+// Returns true of the basis vectors are orthogonal (perpendicular), so it has no skew or shear.
+// See https://en.wikipedia.org/wiki/Orthogonal_basis
 bool Basis::is_orthogonal() const {
-	Basis identity;
-	Basis m = (*this) * transposed();
+	const Vector3 x = get_column(0);
+	const Vector3 y = get_column(1);
+	const Vector3 z = get_column(2);
+	return Math::is_zero_approx(x.dot(y)) && Math::is_zero_approx(x.dot(z)) && Math::is_zero_approx(y.dot(z));
+}
 
-	return m.is_equal_approx(identity);
+// Returns true if the basis vectors are orthonormal (orthogonal and normalized), so it has no scale, skew, or shear.
+// See https://en.wikipedia.org/wiki/Orthonormal_basis
+bool Basis::is_orthonormal() const {
+	const Vector3 x = get_column(0);
+	const Vector3 y = get_column(1);
+	const Vector3 z = get_column(2);
+	return Math::is_equal_approx(x.length_squared(), 1) && Math::is_equal_approx(y.length_squared(), 1) && Math::is_equal_approx(z.length_squared(), 1) && Math::is_zero_approx(x.dot(y)) && Math::is_zero_approx(x.dot(z)) && Math::is_zero_approx(y.dot(z));
 }
```

Comment:
What are the pros and cons of doing this versus the old transpose + identity check? Is one faster than the other / handle more cases etc? :thinking:

Reply:
I need a real is_orthgonal method in the GLTF code. That's why I opened this PR.

Reply:
@nlupugla Shear/skew matrices cannot be decomposed as TRS. This is why skew was added as property in 2D.

Reply:
Hmm, I'm not sure that's true, at least if my understand if TRS is correct/matches yours. Can you give me an example or link me to where this discussion took place?

Here is a worked example decomposing a particular shear matrix into a rotation times a scale. Could it be that for you, a scale matrix is necessarily diagonal? To me, a scale matrix just has to be diagonalizable (ie: diagonal in some orthonormal basis).

<img width="392" alt="image" src="https://github.com/godotengine/godot/assets/12731832/85d89161-8840-4454-9332-08d63fb694b4">

Reply:
I mostly work in 3d. You can do skew and shear by having two TRS applied with non-uniform scale in 3d.

Reply:
I'm not sure what you mean by "diagonal in some orthonormal basis". Let's go with a simple example:

![shear](https://github.com/godotengine/godot/assets/1646875/6ac6612f-5f3f-41dd-9d7f-2da9d1d59b02)

This cannot be decomposed into rotation and scale.

Proof by degrees of freedom: In 2D, rotation is a single angle number, and scale is 2 numbers, so a total of 3 numbers. A 2x2 matrix has 4 numbers. So there is necessarily one degree of freedom that cannot be represented by rotation and scale.

Reply:
How things relate to the monitor or computer graphics is not the problem here. The problem is with what we are calling rotation and scale.

You are thinking of scale as a diagonal matrix, but I am thinking of scale as a Vector3. The other values aren't 0, they don't exist. You can convert a Vector3 scale into a diagonal matrix, but a scale is not the same as a scale matrix, at least this is how it works in game engines. If a matrix can't be losslessly converted to a scale Vector3 (except for floating point error) then it's not a pure scale matrix and therefore the matrix contains more than just scale.

Similarly, you are thinking of rotation as a matrix. A rotation can be represented as Euler angles, a normalized Quaternion, and a matrix, and you can convert between these, but a rotation is not the same as a rotation matrix. If a matrix can't be losslessly converted to Euler angles (except for floating point error) then it's not a pure rotation matrix and therefore the matrix contains more than just rotation.

Reply:
Looking around a bit, it does seem there are at least a few contexts where "scaling" does seem to refer to scaling the x, y, and z axes specifically. I'm kind of annoyed people take it to mean that, but I'll defer to that definition if that is how people commonly understand it, especially in game engines.

As for why I'm annoyed, it's because I would consider a transformation like this:

<img width="771" alt="image" src="https://github.com/godotengine/godot/assets/12731832/18d0709f-d6c8-4738-a9ce-5308365a5a25">

a "scale" because you obtain the transformation by stretching everything aligned with the 45 degree axis by a factor of 2.

Personally, I would think of the "scale" of a transformation as the eigenvalues of its corresponding transformation matrix. That said, it's not a hill I'm willing to die on :)